Optimize code

harshil21 opened the pull request:
Optimizations:

- [x] Quiz scraper v2. Now scrapes subscripts and superscripts! Code still messy cause of bad HTML design in web page I think 
:(

- [x]  Now using spaCy instead of TextBlob. Good performance gain achieved. (Will migrate fully to spaCy in future, replacing even Chatterbot). 

- [x] Few lists converted to tuples, unnecessary variables removed, etc.

...looking for other places where logic/code can be improved

Etc.

General changes:
- Removed bad words which weren't bad
- Quiz stuff for group is now persistent between bot restarts (becomes easier for me, and easier to trace potential bugs/errors)
- Bot now greets new members who joined the TMS group with a link to [Introducing: Telegram!](https://t.me/IntroducingTelegram) for users who are using Telegram for the first time.

Issues fixed:

- [x] Closed #28

harshil21 committed:
fixed #28, tried to optimize things

Attempts made at reducing memory usage by deleting variables and using tuples and/or sets instead of lists.

Also removed some bad words which weren't bad.

harshil21 committed:
Use spaCy instead of TextBlob, quiz scraping improvements

- Completely migrated to spaCy. This significantly increases speed.

There are plans to eventually phase out chatterbot and its dependencies and use spaCy/PyTorch(?)/TensorFlow for ML and Text generation instead.

- Use JaccardSimilarity instead of Levenshtein distance in Chatterbot (better results imo).

- Commented out `train_with()` since we don't use it and don't want it to take memory.

- Reworked `quiz_scraper.py`. It now includes formatting such as subscripts and superscripts.

It also now correctly filters questions and choices which don't meet Telegram's polls' character and choices limit.

- Made quiz related info persist between bot restarts

- Added chat_name to chat_data.

harshil21 committed:
update requirements, fix time in morning quote

harshil21 committed:
latest data

harshil21 committed:
improve regex

now removes space before all contractions like he 's, ca n't, etc.

unclesam79 commented on bot_funcs/commands.py:
```diff
@@ -112,7 +110,7 @@ def swear(update: Update, context: CallbackContext) -> None:
 
         while True:
-            swears = r.choices(prohibited, k=4)  # Returns a list of 4 elements
+            swears = r.choices(tuple(prohibited), k=4)  # Returns a list of 4 elements
```
why

harshil21 replied:
Tuple search faster than list search

unclesam79 commented on chatbot.py:
```diff
@@ -1,56 +1,58 @@
-    end = time.time()
-    time_taken = end - start
-    print(f"\n\nThe Shani Sir chatbot has been trained using the corpus {corpus}. Time taken: {time_taken}s")
+# def train_with(corpus: str) -> None:
```
Uncomment

harshil21 replied:
Not using this function, taking up memory at run time

unclesam79 replied:
> Not using this function, taking up memory at run time

It's just the function name that's stored; nothing else is executed

harshil21 replied:
But why store it in the first place?

unclesam79 replied:
Because we use it to train?

harshil21 replied:
No, never used in the code at all. Was only used for the first time we started using the bot.

unclesam79 replied:
Yea man if I ever need to retrain, I need this code

Let it be you loss it doesn't make a difference to memory use at all

harshil21 replied:
Hence, you can simply uncomment if you want to use it. Anyway you'll have to manually open this file and run, if you'll ever train with this. We're going to stop using chatterbot in the future anyway

unclesam79 replied:
I'm tellin ya it makes no reasonable difference to memory usage whether or not it's commented out, but okay.

harshil21 replied:
Probably takes up ~8KB of memory

unclesam79 replied:
Insignificant

unclesam79 commented on files/lad_words.txt:
```diff
@@ -23,22 +22,15 @@ a_s_s
 alabama hot pocket
 alaskan pipeline
 alla
-allah
```
Undo deletionz

harshil21 replied:
But these words are not bad

unclesam79 replied:
> But these words are not bad

It's not a list of bad words.

harshil21 replied:
Those words are not lad

unclesam79 replied:
But this list is where they belong

harshil21 replied:
But it's racist

unclesam79 replied:
How so?

harshil21 replied:
There are words like 'black' and 'christ' which shouldn't be included in filter, might trigger some people

unclesam79 replied:
> There are words like 'black' and 'christ' which shouldn't be included in filter, might trigger some people

Fair point. But then we don't really have a way to check all 2500+ words for that. Wat to do about that?

harshil21 replied:
We do as much as we can, I already scanned through them once or twice

unclesam79 replied:
Okay cool

unclesam79 commented on main.py:
```diff
@@ -109,9 +109,11 @@ def data_view() -> None:
 
 updater.job_queue.run_repeating(bday_wisher.wish, 86400, first=1)  # Runs every time script is started, and once a day.
-updater.job_queue.run_repeating(morning_wisher.morning_goodness, 86400, first=1)
+updater.job_queue.run_repeating(morning_wisher.morning_goodness, 60**2, first=1)
```
just put 3600 loser

harshil21 replied:
Ok

unclesam79 commented on online/quiz_scraper.py:
```diff
@@ -9,67 +9,92 @@
+        for op in ops:  # Loop through choices if extra formatting such as subscripts or superscripts are present
+
+            sub_sup = re.sub('([0-9][)] )*', '', op.next_element.string).strip()  # Remove choice numbers like '1 )'
+            if op.name == 'sup':  # If superscript present, add that
```
```suggestion
            if op.name in ['sup', 'sub']:  # In case of superscripts/subscripts
```

combine the `if` and the `elif` into one

harshil21 replied:
No the substitution is different
`<sup>` block converts numbers into superscripts, and the `<sub>` block converts them into subscripts.

unclesam79 replied:
Ah sorry I missed that

harshil21 committed:
change from 60**2 to 3600

harshil21 committed:
Merge branch 'master' into optimizations

harshil21 committed:
improve post processing for bot replies

- added new reg exps and improved exisiting ones.

- If there's a question present in output, name will be added to the beginning.

- Changed size of text in leaderboard as there are many participants.

- Also added try except clause for stop quiz since there was an error

- Revert verticalalignment back to center

harshil21 committed:
Update main.py

harshil21 committed:
Merge branch 'optimizations' of https://github.com/tmslads/ShanispeakBot into optimizations

harshil21 committed:
Bot now greets new members in TMS group

- Bot now greets new users with a link to knowing more about Telegram and its features.

- Changed text size in leaderboard so it fits in the bar.

harshil21 committed:
welcome message sent only to non-bots

Also fixed a bug before it could happen

unclesam79 reviewed (approved):
I will miss you, TextBlob.
